Added JobTemplate, a preliminary step for ScheduledJob and Workflow

soltysh opened the pull request:
@sdminonne as promised, sorry it took this long :blush: 
@erictune fyi though it does not have to be in for 1.2

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21675)

<!-- Reviewable:end -->

sdminonne commented:
@soltysh thanks.

erictune commented on pkg/apis/batch/v1/types.go:
```diff
@@ -48,6 +48,24 @@ type JobList struct {
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
+	unversioned.TypeMeta `json:",inline"`
+	v1.ObjectMeta        `json:"metadata,omitempty"`
```
use the same hyperlinks as PodTemplate comments

erictune commented on pkg/apis/extensions/types.go:
```diff
@@ -522,6 +522,24 @@ type JobList struct {
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
+	unversioned.TypeMeta `json:",inline"`
+	api.ObjectMeta       `json:"metadata,omitempty"`
```
same here

erictune commented:
I suggest we wait for after 1.2 to merge this.

erictune commented:
So, this PR seems fine.  

I'm guessing the next PR is going to add "ScheduledJob".  I am not sure where that goes.
I don't think it can go into batch/v1 since `ScheduledJob` won't be `v1` on the first release.

I think what needs to happen is that we need to create a `batch/v2alpha1` which has `Job` and `ScheduledJob` in it, and that group needs to advance through `batch/v2beta1` and then `batch/v2`,
at which point `ScheduledJob` will be GA.  During that time, Jobs will be accessible through 3 or more endpoints.  It seems a bit complex, but I'm not sure what the alternative is.
@lavalamp @bgrant0607 @deads2k 

By the way, it makes it kind of an incentive to get Workflow in at the same time as ScheduledJob, to avoid a `v3alpha1`.

soltysh commented:
> I suggest we wait for after 1.2 to merge this.

Yes, that was the point.

> I think what needs to happen is that we need to create a batch/v2alpha1 which has Job and ScheduledJob in it

Why can't we just use `Job` from `batch/v1` I'd personally stay away from creating Jobs in it. 

Other silly question, instead of messing `batch` sub-group, can't we use `extensions/v1beta1` or `extensions/v2beta1` as the incubator for new types, since the internal is already in place. And then move to appropriate group only when GA?

soltysh commented:
@erictune added the urls you've asked for.

sdminonne commented:
@erictune  @soltysh, I plan to submit a PR for workflow and for the moment it will be in `pkg/apis/extensions/v1beta1` but I'm OK to modify according to other needs/policy

erictune commented:
@bgrant0607 @lavalamp we need to decide what to do in this case:

- we are introducing an alpha-version of a new Kind (e.g. ScheduledJob or Workflow) 
- we know already what the API group for the new kind it will eventually be (e.g. Batch)
- there is already something in the eventual destination group (e.g. Job is already in Batch/v1).

Considerations:
- extensions/v1beta1 is where we have been putting stuff, so it is easy/consistent to put new things there
- but... extensions/v1beta1 was a hack for when we didn't have api groups working in general.  we should stop using it now?
- the internal for Job is in extensions/v1beta1, and so putting things that use it in extensions/v1beta1 is convenient
- but... we have a plan to remove Job from there in version 1.4, and we would not be able to migrate ScheduledJob or Workflow out of extensions on the same schedule
- assuming ScheduledJob or Workflow went into extensions first, and then was ready for GA, where does it go for GA?  Batch/v1 or Batch/v2?  The former makes sense because there is not any change to Job planned, so no need to rev the group for that.  And it is less work.  The latter makes sense since we once said that we would not add new things to an API group once it had been formed, so that users could use the presence of the api group to determine the presence of all enclosing kinds.

bgrant0607 commented:
@erictune If there's nothing here that needs to be decided for 1.2, I'll have to look at this later.

erictune commented:
no 1.2 issues.

lavalamp commented:
Ping me again after 1.2 and I'll have lots to say. :)

On Thu, Feb 25, 2016 at 9:06 AM, Eric Tune <notifications@github.com> wrote:

> no 1.2 issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21675#issuecomment-188880265>
> .
>

sdminonne commented:
@soltysh ping :)

lavalamp commented:
I'm paying attention to PRs again. I will come back to this one once I triage my entire queue.

sdminonne commented:
Thanks

soltysh commented:
@lavalamp yeah, that will be needed for the scheduled jobs I'm hoping to start next week.

lavalamp commented on pkg/apis/batch/v1/types.go:
```diff
@@ -48,6 +48,29 @@ type JobList struct {
+type JobTemplateSpec struct {
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
```
You should not need this except in top-level objects.

soltysh replied:
This is similar to `PodTemplateSpec` to allow us define created Job's metadata fields.

soltysh replied:
I'll update the comment to give my reasoning.

lavalamp commented on pkg/apis/batch/v1/types.go:
```diff
@@ -48,6 +48,29 @@ type JobList struct {
+
+	// Template defines jobs that will be created from this template
+	// http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Template JobTemplateSpec `json:"template,omitempty"`
```
Do you need the JobTemplateSpec at all? Why not just JobSpec?

soltysh replied:
see prev comment for explanation

lavalamp replied:
Ah, of course.

lavalamp commented on pkg/apis/extensions/types.go:
```diff
@@ -525,6 +525,29 @@ type JobList struct {
 }
 
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
```
Don't add JobTemplate here? Do you really need it in both places?

lavalamp replied:
Sorry, this one I understand why you're adding.

lavalamp replied:
(Of course this should get moved to pkg/apis/batch/types.go in the near future, but I guess it's ok here now)

soltysh replied:
That will be done in a separate PR, this PR was initiated by @sdminonne request and the move was mentioned just recently.

lavalamp commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -528,6 +528,24 @@ type JobList struct {
 }
 
+// JobTemplate describes a template for creating copies of a predefined pod.
+type JobTemplate struct {
```
Sorry, I meant don't add it here.

soltysh replied:
Yeah, that makes sense, since we'll be deprecating this endpoint soon.

soltysh commented:
@lavalamp removed from v1beta1 and updated the comments, ptal

lavalamp commented:
@k8s-bot test this issue: #23365

OK, this LGTM.

lavalamp commented:
Welll... I don't know, actually I'm a bit nervous just adding a thing to a GA api. Maybe we should make batch/v2alpha1?

soltysh commented:
> Welll... I don't know, actually I'm a bit nervous just adding a thing to a GA api. Maybe we should make batch/v2alpha1?

@erictune @bgrant0607 opinions?

bgrant0607 commented:
v2alpha1 SGTM

bgrant0607 commented on pkg/apis/batch/v1/types.go:
```diff
@@ -48,6 +48,29 @@ type JobList struct {
+type JobTemplateSpec struct {
+	// Standard object's metadata of the jobs created from this template.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
```
This problem exists in all existing controllers also, but we should really create a stripped-down ObjectMeta to use in templates that only contains the relevant fields. The other fields are a liability to evolution, since the fields will actually be decoded and persisted, and make our documentation misleading and our schema validation incorrect.

soltysh replied:
Moved to https://github.com/kubernetes/kubernetes/issues/23475.

lavalamp commented:
v2alpha1 means:

* make the batch/v2alpha1 directory
* copy everything from v1 to it
* add the new type

Let me know if you encounter any difficulty getting it to work, we haven't
had to do this in a while but we definitely want to do it more in the
future so we need to streamline the process.

On Wed, Mar 23, 2016 at 9:21 AM, Brian Grant <notifications@github.com>
wrote:

> v2alpha1 SGTM
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21675#issuecomment-200420829>
>

soltysh commented:
@lavalamp will do, thx.

erictune commented:
v2alpha1 plan is okay with me.

soltysh commented:
@lavalamp updated, ptal
The only problem I have is locally failing `make verify` with this error:
```
Verifying hack/../hack/verify-godep-licenses.sh
Checking for 'Godeps/' changes against '/job_template'
fatal: Not a valid object name /job_template
FAILED   hack/../hack/verify-godep-licenses.sh	0s
Verifying hack/../hack/verify-godeps.sh
Checking for 'Godeps/' changes against '/job_template'
fatal: Not a valid object name /job_template
FAILED   hack/../hack/verify-godeps.sh	0s
```
@erictune any ideas?

soltysh commented:
@erictune sorry, I've meant @eparis ^

lavalamp commented:
I think something about your local git clone is breaking the script. Could it be that it doesn't like branch names with underscores? It would be astonishing if you're the only person to do that.

lavalamp commented:
The problem that the unit tests have is `pkg/apis/batch/v2alpha1/register.go:21: imported and not used: "k8s.io/kubernetes/pkg/api/v1"`

lavalamp commented on pkg/apis/batch/v2alpha1/register.go:
```diff
@@ -0,0 +1,39 @@
+// Adds the list of known types to api.Scheme.
+func addKnownTypes(scheme *runtime.Scheme) {
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&JobTemplate{},
```
You need to add all of the types here.

lavalamp commented on pkg/apis/batch/v2alpha1/types.go:
```diff
@@ -0,0 +1,46 @@
+	"k8s.io/kubernetes/pkg/api/v1"
+	batchv1 "k8s.io/kubernetes/pkg/apis/batch/v1"
+)
+
```
You need to copy all of the types from v1 here. Each version needs to be a complete set.

lavalamp commented on pkg/apis/batch/v2alpha1/register.go:
```diff
@@ -0,0 +1,44 @@
+	scheme.AddKnownTypes(SchemeGroupVersion,
+		&JobTemplate{},
+		&batchv1.Job{},
+		&batchv1.JobList{},
```
Sorry, don't just reference from batch/v1, actually copy the types into v2alpha1/types.go.

lavalamp commented on pkg/apis/batch/v1/register.go:
```diff
@@ -19,14 +19,12 @@ package v1
 import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apis/batch"
```
Do not introduce this dependency!

lavalamp commented on pkg/apis/batch/v1/register.go:
```diff
@@ -19,14 +19,12 @@ package v1
 )
 
-// GroupName is the group name use in this package
-const GroupName = "batch"
```
It is OK to repeat this because batch/v1 needs to be importable without batch/.

lavalamp commented on hack/test-go.sh:
```diff
@@ -58,7 +58,7 @@ KUBE_GOVERALLS_BIN=${KUBE_GOVERALLS_BIN:-}
 # separated by comma, lists are separated by semicolon. e.g.,
 # "v1,compute/v1alpha1,experimental/v1alpha2;v1,compute/v2,experimental/v1alpha3"
-KUBE_TEST_API_VERSIONS=${KUBE_TEST_API_VERSIONS:-"v1,extensions/v1beta1,metrics/v1alpha1;v1,autoscaling/v1,batch/v1,extensions/v1beta1,metrics/v1alpha1"}
+KUBE_TEST_API_VERSIONS=${KUBE_TEST_API_VERSIONS:-"v1,extensions/v1beta1,metrics/v1alpha1;v1,autoscaling/v1,batch/v1,batch/v2alpha1,extensions/v1beta1,metrics/v1alpha1"}
```
This will slow our tests down even more, but you may need to start a new group, I think this is selecting a version from each group so it will confuse the system to have a group in it twice?

e.g. append something like `;v1,autoscaling/v1,batch/v2alpha1,extensions/v1beta1`

lavalamp commented on hack/test-go.sh:
```diff
@@ -326,7 +326,7 @@ for (( i=0, j=0; ; )); do
   # register the groups/versions as supported by k8s. So KUBE_API_VERSIONS
   # needs to be the superset of KUBE_TEST_API.
-  KUBE_TEST_API="${apiVersion}" KUBE_API_VERSIONS="v1,autoscaling/v1,batch/v1,extensions/v1beta1,componentconfig/v1alpha1,metrics/v1alpha1,authorization.k8s.io/v1beta1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
+  KUBE_TEST_API="${apiVersion}" KUBE_API_VERSIONS="v1,autoscaling/v1,batch/v1,batch/v2alpha1,extensions/v1beta1,componentconfig/v1alpha1,metrics/v1alpha1,authorization.k8s.io/v1beta1" ETCD_PREFIX=${etcdPrefix} runTests "$@"
```
I think it's ok to have both in this list. @nikhiljindal to confirm.

nikhiljindal replied:
Yes should be fine.
We will get to know that its not fine if the tests fail :)

soltysh commented:
I've copied entire `types.go` into `v2alpha1`, but I'm still missing some generations apparently, since tests are failing. Working on it...

soltysh commented:
@lavalamp the linked issue might be fixed with this PR https://github.com/kubernetes/kubernetes/pull/23208 by @deads2k. It would definitely be beneficial to have that one first and only then come back to this one. I'll be looking into that other PR, appreciate your time as well on that PR.

erictune commented:
@lavalamp @soltysh Are you saying that this PR is blocked on  #23208?

deads2k commented:
> @lavalamp @soltysh Are you saying that this PR is blocked on #23208?

From what I've gathered, its a question of enablement.  It's currently hard to reason about which versions of which APIs are enabled, which storage encoding they're using, and which storage destination is in play.  Being unable to reason about that is making it hard to work out why multiple versions of the same API group don't seem to be starting correctly

erictune commented:
@lavalamp @soltysh how is this going?  Can I do anything to unblock it or otherwise help progress on ScheduledJob?

soltysh commented:
@erictune sorry, was busy closing openshift 1.2. I'm back to the topic. I'm updating the PR and will ask for help where needed. What's currently block is the linked issue. I'll try to rebase my PR on top of @deads2k changes.

soltysh commented:
I've rebased and updated the PR with remaining conversions fixes (see last commit for that). The last problem I'm struggling with is clients, iow. this error in unit tests:
```
--- FAIL: TestListJob (0.00s)
...
received: "/apis/batch/v1/jobs", expected: "/apis/batch/v2alpha1/jobs"
```
From looking into client code [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/jobs.go#L106) and [here](https://github.com/kubernetes/kubernetes/blob/31de62216dea226f0553e698da38000b73684d49/pkg/client/unversioned/batch.go#L36) our unversioned client is not that much unversioned. The question is how to add `batch/v2alpha1` there?
@erictune @lavalamp ideas?

lavalamp commented:
I added a line to the test framework to let you fix that error, IIRC. ... Yeah, it was for that test, too.

This resource/version/group: https://github.com/kubernetes/kubernetes/blob/dae5ac482861382e18b1e7b2943b1b7f333c6a2a/pkg/client/unversioned/jobs_test.go#L106

Needs to match the path constructed here: https://github.com/kubernetes/kubernetes/blob/dae5ac482861382e18b1e7b2943b1b7f333c6a2a/pkg/client/unversioned/jobs_test.go#L88

Right now the test is called twice. Now you should call it three times, for extensions/v1beta1, for batch/v1, and batch/v2alpha1.

soltysh commented:
@lavalamp thanks will update shortly

soltysh committed:
Added JobTemplate, a preliminary step for ScheduledJob and Workflow

soltysh committed:
Generated changes for batch/v2alpha1

soltysh commented:
I've rebase this PR on top of #23493 since it's almost in. One remaining thing, I'm working on right now, is failing tests in `pkg/master`. Hopefully I'll be able to nail that down tomorrow.

soltysh committed:
extensions -> batch internals move leftovers

soltysh commented:
@caesarxuchao I found one leftover from #23493, please have a look at [first commit](https://github.com/kubernetes/kubernetes/pull/21675/commits/102b94a33eee636e0b37e6784ffcfd7b155ccf56).

@lavalamp I've slightly extended `testapi.go` (see [last commit](https://github.com/kubernetes/kubernetes/pull/21675/commits/dca4d75c0b793c17923a9461319de19844c5eb95)) to allow having multiple versions at once, this works, but to have fully testable multiple versions (the one in `pkg/client/unversioned/jobs_test.go` you've pointed me to) I need to modify `simple_testclient.go` to allow choosing which version to go with, see this [TODO](https://github.com/kubernetes/kubernetes/blob/d609e6cc98ab27acbe7a7e4101a1d9fe65e354ec/pkg/client/unversioned/testclient/simple/simple_testclient.go#L90-L91) (@caesarxuchao any idea how to fix it?). 
I've started working on the fixes for that by extending [ResourcePath](https://github.com/kubernetes/kubernetes/blob/d609e6cc98ab27acbe7a7e4101a1d9fe65e354ec/pkg/api/testapi/testapi.go#L241) and [ResourcePathWithPrefix](https://github.com/kubernetes/kubernetes/blob/d609e6cc98ab27acbe7a7e4101a1d9fe65e354ec/pkg/api/testapi/testapi.go#L211) to pass which version you're specifically want to go with, but I'm not sure that's the right approach, I'd rather submit a separate PR addressing that problem, if you don't mind. This PR seems to be already big enough.

caesarxuchao commented:
> @caesarxuchao I found one leftover from #23493, please have a look at first commit.

That LGTM.

> I need to modify simple_testclient.go to allow choosing which version to go with, see this TODO (@caesarxuchao any idea how to fix it?). 

IIUC, you want to test multiple versions of the batch group? I think we need to support that in the real client as well, not in simple_testclient.go. IIRC, what we used to do was initializing multiple clients for each version.

lavalamp commented:
Re: testing multiple versions, there are some ENV vars in the test scripts
that let you run the tests multiple times with various groups turned on. We
have two iterations right now, may need to add a 3rd.

On Tue, Apr 26, 2016 at 11:52 AM, Chao Xu <notifications@github.com> wrote:

> @caesarxuchao <https://github.com/caesarxuchao> I found one leftover from
> #23493 <https://github.com/kubernetes/kubernetes/pull/23493>, please have
> a look at first commit.
>
> That LGTM.
>
> I need to modify simple_testclient.go to allow choosing which version to
> go with, see this TODO (@caesarxuchao <https://github.com/caesarxuchao>
> any idea how to fix it?).
>
> IIUC, you want to test multiple versions of the batch group? I think we
> need to support that in the real client as well, not in
> simple_testclient.go. IIRC, what we used to do was initializing multiple
> clients for each version.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21675#issuecomment-214848002>
>

caesarxuchao commented:
The ENV lavalamp is talking about: https://github.com/kubernetes/kubernetes/blob/945daaae91722d0fe59adb689789c20966544ece/hack/test-go.sh#L61

soltysh commented:
@caesarxuchao make sense, I was a little bit confused after talking to @lavalamp yesterday when he mentioned having all versions (iow. both `v1` and `v2alpha1`) enabled at the same time. I'll update the PR with that limitation in mind, at least for now.

lavalamp commented:
sorry for not being clear-- I'm kinda buried right now. If it's easier,
feel free to just turn on all three at the same time, using that env var.

On Tue, Apr 26, 2016 at 1:39 PM, Maciej Szulik <notifications@github.com>
wrote:

> @caesarxuchao <https://github.com/caesarxuchao> make sense, I was a
> little bit confused after talking to @lavalamp
> <https://github.com/lavalamp> yesterday when he mentioned having all
> versions (iow. both v1 and v2alpha1) enabled at the same time. I'll
> update the PR with that limitation in mind, at least for now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21675#issuecomment-214878830>
>

soltysh commented:
No worries, I'll revisit the approach tomorrow morning and would like to push final version tomorrow. I hope with all the bits around versioning settled I'll be able to do so.

soltysh committed:
enable batch/v2alpha1 tests

soltysh commented:
@erictune @lavalamp this is ready for final review. 
A couple of takeaways that needs addressing, but in separate PRs:
- currently we enable all versions at once from a single group, this is because our `install.go` is at a group level and is installing the entire group with all its versions, it needs to be more granular, imho. This is why I've added `batch/v2alpha1` to the same testing group as `batch/v1`.
- we need multiversion client support, to enable testing different versions at once, not possible yet, eg. in `jobs_test.go` only `extensions/v1beta1` and `batch/v1` is being tested. Some small part were addressed to allow `master_test.go` to pass, see the last commit (enable batch/v2alpha1 tests).

soltysh commented:
I was missing newline at the end of `pkg/apis/batch/v2alpha1/generated.proto`.

erictune commented:
@lavalamp this PR is at that magical moment when all its tests are passing and it does not need a rebase.  It just needs an LGTM.  If you could LGTM soon that would be awesome.

lavalamp commented on pkg/master/master.go:
```diff
@@ -316,7 +319,8 @@ func (m *Master) InstallAPIs(c *Config) {
 			VersionedResourcesStorageMap: map[string]map[string]rest.Storage{
-				"v1": batchResources,
+				"v1":       batchv1Resources,
+				"v2alpha1": batchv2alpha1Resources,
```
I think you should probably check that the version is enabled before adding to this map, so like construct the map here and have two `if c.APIResourceConfigSource.AnyResourcesForVersionEnabled(batchapiv2alpha1.SchemeGroupVersion) { apiGroupInfo.VersionedResourcesStorageMap["v2alpha1"] = batchv2alpha1Resources }` statements afterwards.

I don't think it's that big of a deal but without this there's no way to e.g. turn the alpha version off and leave the other version on.

soltysh replied:
Done.

lavalamp commented on pkg/apis/batch/v2alpha1/conversion.go:
```diff
@@ -0,0 +1,143 @@
+	} else {
+		out.ActiveDeadlineSeconds = nil
+	}
+	// unable to generate simple pointer conversion for unversioned.LabelSelector -> v2alpha1.LabelSelector
```
I wasn't going to say anything, but since I already asked for a change, does it work if you just provide the Convert_unversioned_LabelSelector_To_v2alpha1_LabelSelector function? I think the autogenerated conversion functions should call that if you provide it, and then you wouldn't need this function.

soltysh replied:
I've tried it, but although it didn't generate `Convert_unversioned_LabelSelector_To_v2alpha1_LabelSelector` (or the other side). It still generate `autoConvert_unversioned_LabelSelector_To_v2alpha1_LabelSelector`. Which I've decided will be unnecessary code duplication. I'll leave this as is in that case.

lavalamp commented on pkg/api/testapi/testapi.go:
```diff
@@ -55,9 +55,10 @@ var (
-	externalGroupVersion unversioned.GroupVersion
-	internalGroupVersion unversioned.GroupVersion
-	internalTypes        map[string]reflect.Type
+	// the first element in the group is meant to be the preferred version
```
Thanks for dealing with this file. I think it needs some help in the future but I won't ask you do to that.

soltysh replied:
Yes it does :/. My work is just a tip of an iceberg we need to dig into to get multiple versions fully testable.

lavalamp commented:
It looks great, just one minor thing.

soltysh commented:
@lavalamp I've rebased the PR and addressed that comment regarding turning on the storage. Additionally, I've included the swagger spec leftovers I've found when working on scheduled job (see [this commit](https://github.com/kubernetes/kubernetes/pull/24970/commits/613f3e293c849fa128bd3cee3236464ae4b87a4e)) for that particular changes I've included here. ptal

lavalamp commented:
LGTM, thanks!!

erictune commented:
This is so close to the head of the submit queue.  🍀 🙏

soltysh commented:
Rebased. @erictune || @lavalamp mind re-adding the lgtm label.

soltysh commented:
Sorry, forgot to run `hack/update-all.sh`.

soltysh commented:
Rebased, I didn't forget to run `hack/update-all.sh` this time, so hopefully should be a green one. I've also updated `defaults.go` according to recent changes. 
@erictune || @lavalamp mind re-adding the lgtm label.

soltysh commented:
@erictune || @lavalamp mind re-adding the lgtm label, I'd like to avoid rebasing once again.

soltysh commented:
@k8s-bot test this github issue: #24211

soltysh commented:
:tada:

erictune commented:
👍
